File: 152/maximum-product-subarray.go
package main

import "fmt"

func maxProduct(nums []int) int {
	neg, pos := 1, 1
	res := -(1 << 32)
	var negMul, posMul int
	for _, num := range nums {
		negMul, posMul = neg*num, pos*num
		neg = min(negMul, posMul, 1)
		pos = max(posMul, negMul, 1)
		res = max(negMul, posMul, res)
	}
	return res
}

func max(a, b, c int) int {
	res := a
	if res < b {
		res = b
	}
	if res < c {
		res = c
	}
	return res
}

func min(a, b, c int) int {
	res := a
	if res > b {
		res = b
	}
	if res > c {
		res = c
	}
	return res
}

func main() {
	nums := []int{-2, 0, -1}
	fmt.Println(maxProduct(nums))
}

File: 152/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        neg, pos = 1, 1
        res = - (1 << 32)
        for num in nums:
            neg_mul, pos_mul = neg * num, pos * num
            neg = min(neg_mul, pos_mul, 1)
            pos = max(pos_mul, neg_mul, 1)
            res = max(neg_mul, pos_mul, res)
        return res


nums = [-2, 0, -1]
print(Solution().maxProduct(nums))